File: pc/src/valerie/provider/LocalImdbProvider.java
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */

package valerie.provider;

import java.net.URL;
import java.util.ArrayList;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import valerie.MediaInfo;
import valerie.tools.DebugOutput;

/**
 *
 * @author i7
 */
public class LocalImdbProvider {
    private static final String URL = "http://www.imdb.<lang>/";
    private static final String apiSearch = URL + "find?q=<search>";
    private static final String apiPlot = URL + "title/<imdbid>/plotsummary";

    private static final String apiEpisodeList = URL + "title/<imdbid>/episodes";

    private static final String NO_PLOT_RESULT = "id=\"swiki_empty\"";

    private static final String DIV_TITLE_START_EXAMPLE = "<a class=\"main\" href=\"/title/tt0416449/\">";
    private static final String DIV_TITLE_START = "<a class=\"main\" href=\"/title/tt";
    private static final String DIV_TITLE_END = "</a>";

    private static final String DIV_PLOT_START = "<div id=\"swiki.2.1\">";
    private static final String DIV_PLOT_END = "</div>";


    private static class ResultEntry {
        public String  ImdbId;
        public String  Title;
        public int Season;
        public int Episode;

        public ResultEntry() {
            Season = -1;
            Episode = -1;
        }

        @Override
        public String toString() {
            return Title + ":" + Season + ":" + Episode + ":" + ImdbId;
        }
    }

    private static final String DIV_EPISODE_START = "<tr> <td valign=\"top\"><h3>";
    private static final String DIV_EPISODE_FLAG = ">";
    private static final String DIV_EPISODE_END = "</a></h3>";
    private static ArrayList<ResultEntry> getEpisodes(String html, String lang) {
        ArrayList<ResultEntry> results = new ArrayList<ResultEntry>();

        String[] htmlSplitted = html.split(DIV_EPISODE_START);
        for(String htmlSplitter : htmlSplitted) {
            htmlSplitter = htmlSplitter.trim();
            if(!htmlSplitter.startsWith(DIV_EPISODE_FLAG))
                continue;

            int pos = htmlSplitter.indexOf(DIV_EPISODE_END);
            if(pos < 0)
                continue;

            ResultEntry entry = new ResultEntry();
            String strEntry = htmlSplitter.substring(0, pos);

            Pattern pImdbId = Pattern.compile("/title/tt\\d*/");
            Matcher mImdbId = pImdbId.matcher(strEntry);
            if(mImdbId.find()) {
                String sImdbId = mImdbId.group();
                sImdbId = sImdbId.replaceAll("/title/", "");
                sImdbId = sImdbId.replaceAll("/", "");
                entry.ImdbId = sImdbId;
            }


            String season_start = "";
            String season_end = "";
            String episode_start = "";
            String episode_end = "";

            if(lang.equals("de")) {
                season_start = "Staffel ";
                season_end = ", ";
                episode_start = "Folge ";
                episode_end = ": ";
            } else if(lang.equals("it")) {
                season_start = "Stagione ";
                season_end = ", ";
                episode_start = "Episodio ";
                episode_end = ": ";
            } else if(lang.equals("es")) {
                season_start = "Temporada ";
                season_end = ", ";
                episode_start = "Episodio ";
                episode_end = ": ";
            } else if(lang.equals("fr")) {
                season_start = "Saison ";
                season_end = ", ";
                episode_start = "Episode ";
                episode_end = ": ";
            } else if(lang.equals("pt")) {
                season_start = "Temporada ";
                season_end = ", ";
                episode_start = "Epis&#xF3;dio "; //TODO: I dont belive that this will work as this should be already converted to UTF-8
                episode_end = ": ";
            }

            pos = strEntry.indexOf(season_start);
            if(pos >= 0) {
                String se = strEntry.substring(pos + season_start.length());
                pos = se.indexOf(season_end);
                if(pos >= 0) {
                    String season = se.substring(0, pos);
                    entry.Season = Integer.valueOf(season);


                }
            }

            pos = strEntry.indexOf(episode_start);
            if(pos >= 0) {
                String se = strEntry.substring(pos + episode_start.length());
                pos = se.indexOf(episode_end);
                if(pos >= 0) {
                    String episode = se.substring(0, pos);
                    entry.Episode = Integer.valueOf(episode);


                }
            }

            results.add(entry);
        }

        return results;
    }

    private static boolean getTitle(MediaInfo info, String html) {
        //<a class="main" href="/title/tt0416449/">300</a>
        String title = html;
        //if(title.contains(NO_PLOT_RESULT)) {
        //    return false;
        //}

        int pos = title.indexOf(DIV_TITLE_START);
        if(pos < 0)
            return false;

        title = title.substring(pos + DIV_TITLE_START_EXAMPLE.length());

        pos = title.indexOf(DIV_TITLE_END);
        if(pos < 0)
            return false;

        info.Title = title.substring(0, pos).replaceAll("\"", "").trim();

        ////

        title = html;
        pos = title.indexOf(DIV_TITLE_START, pos);
        if(pos < 0)
            return true;

        title = title.substring(pos + DIV_TITLE_START_EXAMPLE.length());

        pos = title.indexOf(DIV_TITLE_END);
        if(pos < 0)
            return true;

        info.Title = title.substring(0, pos).replaceAll("\"", "").trim();

        return true;
    }

    private static boolean getPlot(MediaInfo info, String html) {
        String plot = html;
        if(plot.contains(NO_PLOT_RESULT)) {
            return false;
        }

        int pos = plot.indexOf(DIV_PLOT_START);
        if(pos < 0)
            return false;

        plot = plot.substring(pos + DIV_PLOT_START.length());

        pos = plot.indexOf(DIV_PLOT_END);
        if(pos < 0)
            return false;

        info.Plot = plot.substring(0, pos).replaceAll("<br>", " ").trim() + " [IMDB.LOCAL]";
        return true;
    }


    public static boolean getMoviesByImdbID(MediaInfo info, String lang) {
        return getMoviesByImdbID(info, lang, null);
    }

    private static boolean getMoviesByImdbID(MediaInfo info, String lang, String imdbid) {

        DebugOutput.printl("");

        String html = null;

        String url = apiPlot;
        if(imdbid != null)
            url = url.replaceAll("<imdbid>", imdbid);
        else
            url = url.replaceAll("<imdbid>", info.ImdbId);
        url = url.replaceAll("<lang>", lang);
        html = valerie.tools.WebGrabber.getHtml(url);

        if (html == null)
            return false;

        getTitle(info, html);
        getPlot(info, html);

        return true;
    }

    public static boolean getEpisodeByImdbID(MediaInfo info, String lang) {

        DebugOutput.printl("");

        if(info.Episode == -1 || info.Season == -1)
            return false;

        String html = null;

        String url = apiEpisodeList;
        url = url.replaceAll("<imdbid>", info.ImdbId);
        url = url.replaceAll("<lang>", lang);
        html = valerie.tools.WebGrabber.getHtml(url);

        if (html == null)
            return false;

        ArrayList<ResultEntry> results = getEpisodes(html, lang);
        for(ResultEntry result : results) {
            if(result.Season == info.Season && result.Episode == info.Episode) {
                return getMoviesByImdbID(info, lang, result.ImdbId);
            }
        }

        return false;
    }


}

File: pyvalerie/src/LocalImdbProvider.py
'''
Created on 29.11.2010

@author: i7
'''

import re
import WebGrabber
import Utf8

class LocalImdbProvider():
    '''
    classdocs
    '''

    URL = u"http://www.imdb.<lang>/"
    apiSearch = URL + u"find?q=<search>"
    apiPlot   = URL + u"title/<imdbid>/plotsummary"

    apiEpisodeList = URL + u"title/<imdbid>/episodes"

    NO_PLOT_RESULT = u"id=\"swiki_empty\""

    DIV_TITLE_START_EXAMPLE = u"<a class=\"main\" href=\"/title/tt0416449/\">"
    DIV_TITLE_START = u"<a class=\"main\" href=\"/title/tt"
    DIV_TITLE_END   = u"</a>"

    DIV_PLOT_START = u"<div id=\"swiki.2.1\">"
    DIV_PLOT_END   = u"</div>"
        
    class ResultEntry:
        ImdbId = ""
        Title = ""
        Season = -1
        Episode = -1

        def __init__(self):
            self.Season = -1
            self.Episode = -1

        def __str__(self):
            return Utf8.utf8ToLatin(self.Title + u":" + self.Season + u":" + self.Episode + u":" + self.ImdbId)
        
    DIV_EPISODE_START = u"<tr> <td valign=\"top\"><h3>";
    DIV_EPISODE_FLAG  = u">";
    DIV_EPISODE_END   = u"</a></h3>";
    
    def getResults(self, html, lang):
        results = []
        
        htmlSplitted = html.split(self.DIV_EPISODE_START)
        for htmlSplitter in htmlSplitted:
            htmlSplitter = htmlSplitter.strip()
            if htmlSplitter.startswith(self.DIV_EPISODE_FLAG) is False:
                continue
            
            pos = htmlSplitter.find(self.DIV_EPISODE_END)
            if pos < 0:
                continue
            
            entry = self.ResultEntry()
            strEntry = htmlSplitter[0:pos]
            
            mImdbId = re.search(r'/title/tt\\d*/', strEntry)
            if mImdbId and mImdbId.group():
                sImdbId = mImdbId.group()
                sImdbId = re.sub("/title/", "", sImdbId)
                sImdbId = re.sub("/", "", sImdbId)
                entry.ImdbId = sImdbId

            season_start  = u""
            season_end    = u""
            episode_start = u""
            episode_end   = u""

            if lang == u"de":
                season_start  = u"Staffel "
                season_end    = u", "
                episode_start = u"Folge "
                episode_end   = u": "
            elif lang == u"it":
                season_start  = u"Stagione "
                season_end    = u", "
                episode_start = u"Episodio "
                episode_end   = u": "
            elif lang == u"es":
                season_start  = u"Temporada "
                season_end    = u", "
                episode_start = u"Episodio "
                episode_end   = u": "
            elif lang == u"fr":
                season_start  = u"Saison "
                season_end    = u", "
                episode_start = u"Episode "
                episode_end   = u": "
            elif lang == u"pt":
                season_start  = u"Temporada "
                season_end    = u", "
                episode_start = u"Epis&#xF3;dio " #TODO: I dont belive that this will work as this should be already converted to UTF-8
                episode_end   = u": "

            pos = strEntry.find(season_start);
            if pos >= 0:
                se = strEntry[pos + len(season_start):]
                pos = se.find(season_end);
                if pos >= 0:
                    season = se[0:pos]
                    entry.Season = int(season);


            pos = strEntry.find(episode_start)
            if pos >= 0:
                se = strEntry[pos + len(episode_start):]
                pos = se.find(episode_end)
                if pos >= 0:
                    episode = se[0:pos]
                    entry.Episode = int(episode)

            results.append(entry)
        return results
    
    def getTitle(self, info, html):
        #<a class="main" href="/title/tt0416449/">300</a>
        title = html;
        #if self.NO_PLOT_RESULT in title:
        #    return false;

        pos = title.find(self.DIV_TITLE_START)
        if pos < 0:
            return None

        title = title[pos + len(self.DIV_TITLE_START_EXAMPLE):]

        pos = title.find(self.DIV_TITLE_END)
        if pos < 0:
            return None

        info.Title = title[0:pos].strip()
        info.Title = re.sub("\"", "", info.Title)

        ####

        title = html;
        pos = title.find(self.DIV_TITLE_START, pos)
        if pos < 0:
            return info

        title = title[pos + len(self.DIV_TITLE_START_EXAMPLE):]

        pos = title.find(self.DIV_TITLE_END)
        if pos < 0:
            return info

        info.Title = title[0:pos].strip()
        info.Title = re.sub("\"", "", info.Title)

        return info
    
    def getTag(self, info, html):
        #<a class="main" href="/title/tt0416449/">300</a>
        title = html;
        #if self.NO_PLOT_RESULT in title:
        #    return false;

        pos = title.find(self.DIV_TITLE_START)
        if pos < 0:
            return None

        title = title[pos + len(self.DIV_TITLE_START_EXAMPLE):]

        pos = title.find(self.DIV_TITLE_END)
        if pos < 0:
            return None

        info.Title = title[0:pos].strip()
        info.Title = re.sub("\"", "", info.Title)

        ####

        title = html;
        pos = title.find(self.DIV_TITLE_START, pos)
        if pos < 0:
            return info

        title = title[pos + len(self.DIV_TITLE_START_EXAMPLE):]

        pos = title.find(self.DIV_TITLE_END)
        if pos < 0:
            return info

        info.Title = title[0:pos].strip()
        info.Title = re.sub("\"", "", info.Title)

        return info
    
    def getPlot(self, info, html):
        plot = html
        if self.NO_PLOT_RESULT in plot:
            return None

        pos = plot.find(self.DIV_PLOT_START)
        if pos < 0:
            return None

        plot = plot[pos + len(self.DIV_PLOT_START):]

        pos = plot.find(self.DIV_PLOT_END)
        if pos < 0:
            return None

        info.Plot = plot[0:pos]
        info.Plot = re.sub("<br>", " ", info.Plot).strip() + " [IMDB.LOCAL]"
        return info
    
    ###############################################
    
    # Not needed as other functions has opt parameter
    #def getMoviesByImdbID(self, info, lang):
    #    return self.getMoviesByImdbID(info, lang, None)

    def getMoviesByImdbID(self, info, lang, imdbid=None):

        url = self.apiPlot
        if imdbid is not None:
            url = re.sub("<imdbid>", imdbid, url)
        else:
            url = re.sub("<imdbid>", info.ImdbId, url)
        url = re.sub("<lang>", lang, url)
        html = WebGrabber.getHtml(url)

        if html is None:
            return None;

        tmp = self.getTitle(info, html)
        if tmp is not None:
            info = tmp
        
        tmp = self.getPlot(info, html)
        if tmp is not None:
            info = tmp        

        return info;

    def getEpisodeByImdbID(self, info, lang):

        if info.Episode == -1 or info.Season == -1:
            return None

        url = self.apiEpisodeList
        url = re.sub("<imdbid>", info.ImdbId, url)
        url = re.sub("<lang>", lang, url)
        html = WebGrabber.getHtml(url)

        if html is None:
            return None

        results = self.getResults(html, lang)
        for result in results:
            if result.Season == info.Season and result.Episode == info.Episode:
                return self.getMoviesByImdbID(info, lang, result.ImdbId)
        
        return None    
    